File: crawler.py
import requests
from bs4 import BeautifulSoup
import json
import undetected_chromedriver as uc
from selenium.webdriver.common.by import By
import time
import traceback
import re


def fetch_video_list(url, max_count=25):
    options = uc.ChromeOptions()
    # options.add_argument('--headless')
    options.add_argument('--disable-gpu')
    options.add_argument('--no-sandbox')
    driver = uc.Chrome(options=options)
    driver.get(url)
    print("[除錯] 已開啟瀏覽器並載入主頁，等待 5 秒...")
    time.sleep(5)
    video_items = driver.find_elements(By.CLASS_NAME, 'video-img-box')
    video_list = []
    for item in video_items[:max_count]:
        try:
            img = item.find_element(By.TAG_NAME, 'img')
            img_url = img.get_attribute('data-src') or img.get_attribute('src')
            title_elem = item.find_element(By.CSS_SELECTOR, '.detail .title a')
            title = title_elem.text.strip()
            detail_url = title_elem.get_attribute('href')
            video_res = fetch_video_detail_m3u8(detail_url)
            video_list.append({
                'img_url': img_url,
                'title': title,
                'detail_url': detail_url,
                'video': video_res.get('m3u8_url'),
                'key_url': video_res.get('key_url')
            })
        except Exception as e:
            print(f'[警告] 解析單一影片資料失敗: {e}')
    driver.quit()
    return video_list


def fetch_video_detail_m3u8(detail_url):
    options = uc.ChromeOptions()
    # options.add_argument('--headless')
    options.add_argument('--disable-gpu')
    options.add_argument('--no-sandbox')
    driver = uc.Chrome(options=options)
    driver.get(detail_url)
    time.sleep(1)
    m3u8_url = None
    key_url = None
    try:
        scripts = driver.find_elements(By.TAG_NAME, 'script')
        for script in scripts:
            text = script.get_attribute('innerHTML')
            if text and '.m3u8' in text:
                match = re.search(r'(https?://[^\'"\s]+\.m3u8)', text)
                if match:
                    m3u8_url = match.group(1)
                    break
        if m3u8_url:
            try:
                res = requests.get(m3u8_url, timeout=8)
                if res.status_code == 200:
                    m3u8_txt = res.text
                    key_match = re.search(r'#EXT-X-KEY:METHOD=AES-128,URI="([^"]+)"', m3u8_txt)
                    if key_match:
                        key_url = key_match.group(1)
                        if not key_url.startswith("http"):
                            from urllib.parse import urljoin
                            key_url = urljoin(m3u8_url, key_url)
            except Exception as err:
                print(f"[錯誤] 讀取 m3u8 金鑰失敗: {err}")
    except Exception as e:
        print(f'[錯誤] 解析 m3u8 失敗: {e}')
    driver.quit()
    return {"m3u8_url": m3u8_url, "key_url": key_url}


if __name__ == '__main__':
    target_url = 'https://jable.tv/categories/chinese-subtitle/'
    video_list = fetch_video_list(target_url)
    with open('video_list.json', 'w', encoding='utf-8') as f:
        json.dump(video_list, f, ensure_ascii=False, indent=2)
    print(f'已寫入 {len(video_list)} 筆影片資訊到 video_list.json')

File: lib/crawler.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:html/parser.dart' as parser;

class VideoInfo {
  final String title;
  final String imgUrl;
  final String detailUrl;
  final String? videoUrl;
  final String? keyUrl;

  VideoInfo({
    required this.title,
    required this.imgUrl,
    required this.detailUrl,
    this.videoUrl,
    this.keyUrl,
  });

  Map<String, dynamic> toJson() => {
        'title': title,
        'img_url': imgUrl,
        'detail_url': detailUrl,
        'video': videoUrl,
        'key_url': keyUrl,
      };

  static VideoInfo fromJson(Map<String, dynamic> json) => VideoInfo(
        title: json['title'] ?? '',
        imgUrl: json['img_url'] ?? '',
        detailUrl: json['detail_url'] ?? '',
        videoUrl: json['video'],
        keyUrl: json['key_url'],
      );
}

Future<List<VideoInfo>> fetchVideoList(String url, {int maxCount = 25}) async {
  final res = await http.get(Uri.parse(url));
  if (res.statusCode != 200) return [];
  final doc = parser.parse(res.body);
  final items = doc.getElementsByClassName('video-img-box');
  final videos = <VideoInfo>[];
  for (final item in items.take(maxCount)) {
    final img = item.querySelector('img');
    final imgUrl = img?.attributes['data-src'] ?? img?.attributes['src'] ?? '';
    final titleElem = item.querySelector('.detail .title a');
    final title = titleElem?.text.trim() ?? '';
    final detailUrl = titleElem?.attributes['href'] ?? '';
    final detail = await fetchVideoDetailM3u8(detailUrl);
    videos.add(VideoInfo(
      title: title,
      imgUrl: imgUrl,
      detailUrl: detailUrl,
      videoUrl: detail['m3u8_url'],
      keyUrl: detail['key_url'],
    ));
  }
  return videos;
}

Future<Map<String, String?>> fetchVideoDetailM3u8(String detailUrl) async {
  final res = await http.get(Uri.parse(detailUrl));
  if (res.statusCode != 200) return {'m3u8_url': null, 'key_url': null};
  final doc = parser.parse(res.body);
  String? m3u8Url;
  String? keyUrl;
  for (final script in doc.getElementsByTagName('script')) {
    final text = script.text;
    final match =
        RegExp(r'(https?:\\/\\/[^\'"\s]+\.m3u8)').firstMatch(text);
    if (match != null) {
      m3u8Url = match.group(1)?.replaceAll('\\/', '/');
      break;
    }
  }
  if (m3u8Url != null) {
    final res2 = await http.get(Uri.parse(m3u8Url));
    if (res2.statusCode == 200) {
      final m3u8Txt = res2.body;
      final keyMatch =
          RegExp(r'#EXT-X-KEY:METHOD=AES-128,URI="([^"]+)"').firstMatch(m3u8Txt);
      if (keyMatch != null) {
        keyUrl = keyMatch.group(1);
        if (keyUrl != null && !keyUrl!.startsWith('http')) {
          keyUrl = Uri.parse(m3u8Url).resolve(keyUrl!).toString();
        }
      }
    }
  }
  return {'m3u8_url': m3u8Url, 'key_url': keyUrl};
}

Future<String> encodeVideoList(List<VideoInfo> videos) async {
  final list = videos.map((v) => v.toJson()).toList();
  return jsonEncode(list);
}
